Feat: data-stream support + logging improvements (#988)
logstash-plugins/logstash-output-elasticsearch: docs/index.asciidoc

For more details, check out the
{logstash-ref}/connecting-to-cloud.html[Logstash-to-Cloud documentation].

[id="plugins-{type}s-{plugin}-data_stream"]

> ===== `data_stream`

Author: these are mostly just :scissors: from the [spec](https://github.com/elastic/logstash/issues/12178) could certainly use more data-stream related chit-chat ... // cc @karenzone

* Value can be any of: `true`, `false` and `auto`
* Default value depends on Logstash version.

Defines whether data will be indexed into an Elasticsearch data stream.
The other `data_stream_*` settings will only be used if this setting is enabled.
Defaults to `false` in Logstash 7.x and `auto` starting in Logstash 8.0.

> The `auto` behavior means that whenever the supplied configuration is compatible

Reviewer: ```suggestion
Logstash handles the output as a data stream when the supplied configuration
is compatible with data streams and this value is set to `auto`.
```

with data streams Logstash will assume to be handling events as if `true` was set.

[id="plugins-{type}s-{plugin}-data_stream_auto_routing"]
===== `data_stream_auto_routing`

* Value type is <<boolean,boolean>>
* Default value is `true`.

Automatically routes events by deriving the data stream name using specific event
fields with the `%{data_stream.type}-%{data_stream.dataset}-%{data_stream.namespace}` format.

If enabled, the `data_stream.*` event fields will take precedence over the
`data_stream_type`, `data_stream_dataset`, and `data_stream_namespace` settings,
but will fall back to them if any of the fields are missing from the event.

[id="plugins-{type}s-{plugin}-data_stream_dataset"]
===== `data_stream_dataset`

* Value type is <<string,string>>
* Default value is `generic`.

The data stream dataset used to construct the data stream at index time.

> 

Reviewer: Are there other valid entries at this time?

Author: Believe anything is valid (as long as it's a valid name) in ES. 
Only the `data_stream_type` is contrained with enumerated values.

[id="plugins-{type}s-{plugin}-data_stream_namespace"]
===== `data_stream_namespace`

* Value type is <<string,string>>
* Default value is `default`.

The data stream namespace used to construct the data stream at index time.

[id="plugins-{type}s-{plugin}-data_stream_sync_fields"]
===== `data_stream_sync_fields`

* Value type is <<boolean,boolean>>
* Default value is `true`

> Automatically syncs the `data_stream.*` event fields if they are missing from the
> event, this ensures fields match the data stream name that events are being sent to.

Reviewer: ```suggestion
Automatically adds and syncs the `data_stream.*` event fields if they are missing from the
event. This ensures that fields match the name of the data stream that is receiving events.
```

> NOTE: when `data_stream_auto_routing` is disabled and existing `data_stream.*` event
> fields do not match the data stream name they will be overwritten (with a warning).

Reviewer: ```suggestion
NOTE: If existing `data_stream.*` event fields do not match the data stream name
and `data_stream_auto_routing` is disabled, the event fields will be
overwritten with a warning.
```

[id="plugins-{type}s-{plugin}-data_stream_type"]
===== `data_stream_type`

* Value type is <<string,string>>
* Default value is `logs`.

The data stream type used to construct the data stream at index time.

> Currently, only `logs` and `metrics` is supported.

Reviewer: ```suggestion
Currently, only `logs` and `metrics`are supported.
```
